### PR TITLE
rados: add a framework for read and write operations (new attempt)

### PR DESCRIPTION
**Pull request by phlogistonjohn**
This is a redo of what was found in #315. Please see below for a more detailed list of what is different from that PR.

The Ceph librados API provides a suite of API calls for interacting with "operations". These are sets of steps that put together perform a batch operation atomically. The API flow goes a bit like (in pseudocode):

```
op = create_op()
assert_version(op, ...)
read_omap(op, ...)
read_data(op, ...)
operate(op)
```

In go-ceph we were already using some of these API calls, but we did not expose the batch nature of them. Rather, we would wrap it all in one Go function and effecitvely hid this. While convenient if all you want to do is read or write the omap or "touch" an object, it prevents the user of go-ceph from using the "full power" of the APIs ceph provides.

My goal with this PR is to establish a basic pattern of use in our Go APIs that reflects the power of the ceph APIs in a way that is fairly idiomatic to Go - much like our general approach elsewhere in the library.

In addition, since these are multiple calls that could be spread out over time the new code must properly manage any C memory that the librados APIs need. Some calls expect a pointer to a an array of strings and it could be hours between the time this memory is allocated and it is used by the operate call. This means part of the design in these patches is to support the use of the C memory but ease the management.

For both read and write operations the code flows as such (more Go-ish pseudo code):

```go
op := CreateWriteOp()
defer op.Release()
op.Create()
op.Write(...)
op.SetOmap(...)
err := op.Operate(ioctx, "myoid", OperationNoFlags)
```

In some cases the steps are expected to acquire data when the operation is performed. The functions that add those steps to the operation will return. The returned "step objects" can be used to get data after the operation:

```go
op := CreateWriteOp()
defer op.Release()
op.AssertExists()
omapGetter1 := op.GetOmapValues(...)
err := op.Operate(ioctx, "myoid", OperationNoFlags)
// simplified! check for errors!
for {
    kv, _ := ommapGetter1.Next()
    if kv == nil {
       break
    }
    // do something with the key-value pair
}
```

In general, each step may return a type with a different set of functions but the idea is that the operation is performed and then the return from the step can be used to access data.

This PR does not contain a complete set of functions for either read ops or write ops. It is just enough (IMO) to demonstrate how future operation steps can be added and convert most of the existing uses of operation calls to
this new approach.

## Major differences from PR #315

* What was called an element is now called a step.
* Tweaked and simplified error type.
* Removed the concept of "reset". It turns out that this idea was based on
  a faulty understanding of the Ceph calls. Because ceph level operations can
  not be "used" more than once for a given object, it made no sense to
  try and support it in go-ceph. This futher simplified the code.
* The patch ordering is similar but I have reorganized it a bit to hopefully
  ease reviewing.
* A few new (very simple) test cases have been added.

## Work (intentionally) left undone

* There are some TODOs left in doc comments for ceph flags. It's probably worth
  calling out the most important of these but this will require reading ceph sources.
* Write operations are more fleshed out than read operation calls. This is due
  to reads being a bit more complex than writes to implement and the desire
  not to boil the ocean with this PR.
* The OperationError type is left fairly simplistic and no "helper" methods
  remain in the code. I expect that there will be a use case for something
  like Unwrap for this type but it can be added later.

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Dear team,
I realize that this is fairly close to release but this is essentially a reworked version of #315. It is my sincere wish that we do our best to try and land this in v0.7. I do realize this is a largish series - but if you have helped review #315 most of it should look familiar. I tried to incorporate as much of the feedback on that PR as possible.

Thanks!

(CC: @ansiwen @agarwal-mudit @nixpanic )

File: rados/omap.go [hunk 1]
```diff
@@ -7,63 +7,234 @@ package rados
...
+	// C returned data:
+	iter C.rados_omap_iter_t
+	more C.uchar
+	rval C.int
+
+	// internal state:
+
+	// canIterate is only set after the operation is performed and is
+	// intended to prevent premature fetching of data
+	canIterate bool
```
Comment (ansiwen) [inline rados/omap.go:125]: `isReady`

**Review by ansiwen — CHANGES_REQUESTED** [review]

**Comment by ansiwen** [comment]
I'm still not done yet, but I found at least one issue.

File: rados/omap.go [hunk 2]
```diff
@@ -7,63 +7,234 @@ package rados
...
-		c_value_ptr := (**C.char)(unsafe.Pointer(uintptr(c_values) + uintptr(i)*ptrSize))
-
-		var c_length C.size_t
-		if len(value) > 0 {
-			*c_value_ptr = (*C.char)(unsafe.Pointer(&value[0]))
-			c_length = C.size_t(len(value))
+		cvp := (**C.char)(unsafe.Pointer(uintptr(cValues) + uintptr(i)*psize))
+		vlen := C.size_t(len(value))
+		if vlen > 0 {
+			*cvp = (*C.char)(unsafe.Pointer(&value[0]))
```
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go:70]: Ok, apparently there is `C.CBytes()` in analogy to `C.CString()`, so you should write:
```go
	cv := C.CBytes(key)
	sos.addRefs(cv)
	*cvp = cv
```

[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:70]: I don't think this is a violation. If it is, its one I'm comfortable making as we have over 60 lines of code in go ceph using this technique and its the only one I know of that avoid excessive copying of data.

Additionally, I tried searching around for an official stance on this but could not find one. I did find references to the technique and none of the core go devs responded with a "don't do that":
https://github.com/golang/go/issues/14210#issuecomment-223169706
https://github.com/golang/go/issues/40636

Are some examples.

File: rados/omap.go [hunk 3]
```diff
@@ -7,63 +7,234 @@ package rados
...
+	psize := uintptr(strpSize)
 	i := 0
 	for key, value := range pairs {
 		// key
-		c_key_ptr := (**C.char)(unsafe.Pointer(uintptr(c_keys) + uintptr(i)*ptrSize))
-		*c_key_ptr = C.CString(key)
-		defer C.free(unsafe.Pointer(*c_key_ptr))
+		ck := C.CString(key)
+		sos.addRefs(unsafe.Pointer(ck))
+		ckp := (**C.char)(unsafe.Pointer(uintptr(cKeys) + uintptr(i)*psize))
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:63]: I'd be happy to do that some day. In fact, I tried once myself before I started on this PR, but a combination of cgo pointer checks and even less readable code blocked my path. I also tried abstracting some of the logic into a helper type but that just ended up creating tons more code with lots of hard to follow indirection.

Maybe future developments in the Go space (generics?) would make this possible in the future. I don't know. So, emphasis on "some day".

File: rados/omap.go [hunk 4]
```diff
@@ -7,63 +7,234 @@ package rados
...
+		pairs:    pairs,
+		cKeys:    (**C.char)(cKeys),
+		cValues:  (**C.char)(cValues),
+		cLengths: (*C.size_t)(cLengths),
+		cNum:     C.size_t(len(pairs)),
+	}
+	sos.addRefs(cKeys, cValues, cLengths)
+
+	psize := uintptr(strpSize)
 	i := 0
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:58]: I'll look into that, thanks.

File: rados/omap.go [hunk 5]
```diff
@@ -7,63 +7,234 @@ package rados
...
+)
 
-	var s C.size_t
+func init() {
 	var c *C.char
-	ptrSize := unsafe.Sizeof(c)
+	strpSize = C.size_t(unsafe.Sizeof(c))
+	var s C.size_t
+	sizeTSize = C.size_t(unsafe.Sizeof(s))
+}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:24]: Well, I was basing what I had on what the code was doing before (incremental steps and all that). I was unaware of `C.sizeof_size_t`. I'm not sure the other line `C.size_t(unsafe.Sizeof(unsafe.Pointer(nil)))` would compile outside of a function call, but I can try it out later.

File: rados/operation.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+// collection.  That's one reason the operation type tracks the steps (the
+// parts that track complex inputs and outputs) so that as long as the op
+// exists it will have a reference to the step, which will have references
+// to the C language types.
+
+type opKind string
+
+const (
+	readOp  opKind = "read"
+	writeOp        = "write"
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:29]: It's a fairly common go idiom. I can change it for clarity.

File: rados/operation.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+)
+
+// OperationError is an error type that may be returned by an Operate call.
+// It captures the error from the operate call itself and any any errors from
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:33]: thanks, will fix!

File: rados/operation.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+// Not all types implementing opStep are expected to need all the functions
+// in the interface. However, for the sake of simplicity on the op side, we use
+// the same interface for all cases and expect those implementing opStep
+// just embed the without* types that provide no-op implementation of
+// functions that make up this interface.
+type opStep interface {
+	// update will be called after the op's Operate function. It can be used
+	// to convert values from C and cache them and/or communicate a failure
+	// of the action associated with the step.
+	update() error
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:73]: The update function will only be called once, but it doesn't have to be idempotent IIRC.
On naming, I read it as "*update* the state of the step" now that the operation has been performed.
The terms obtain or resolve don't seem to work for me, and get just seems backwards.

File: rados/operation.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	// free will be called to free any resources, especially C memory, that
+	// the step is managing.
+	free()
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:76]: No, free should be idempotent. The API can't enforce that though, so its up to the writer of each step to ensure it.
I can update the comment to reflect that.

File: rados/operation.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	refs []unsafe.Pointer
+}
+
+func (w *withRefs) free() {
+	r := w.refs
+	w.refs = nil
+	for i := range r {
+		C.free(r[i])
+	}
+}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:145]: It is probably premature defensive programming but I wanted only one call to free to do anything in the case of a race. The get and clear of w.refs is not atomic though, so its weak. But the cost is small, so I don't know if I want to change it.

File: rados/operation.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+func (w *withRefs) addRefs(ptrs ...unsafe.Pointer) {
+	if w.refs == nil {
+		w.refs = make([]unsafe.Pointer, 0)
+	}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:150]: Sure.

File: rados/operation.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+	steps []opStep
+}
+
+// free will call the free method of all the steps this operation
+// contains.
+func (o *operation) free() {
+	for i := range o.steps {
+		o.steps[i].free()
+	}
+}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:91]: I'll look into it.

File: rados/read_op.go [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+// Release the resources associated with this read operation.
+func (r *ReadOp) Release() {
+	C.rados_release_read_op(r.op)
+	r.op = nil
+	r.free()
+}
+
+// Operate will perform the operation(s).
+func (r *ReadOp) Operate(ioctx *IOContext, oid string, flags OperationFlags) error {
+	if err := ioctx.validate(); err != nil {
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:37]: It can, but there's only so much defensive program even I'm willing to do. I'm perfectly happy with this panicking if the caller passes in nil as ioctx. The validate is there to make sure our internal state (that the caller is not allowed to touch is valid).

File: rados/read_op.go [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+func (r *ReadOp) operateCompat(ioctx *IOContext, oid string) error {
+	switch err := r.Operate(ioctx, oid, OperationNoFlag).(type) {
+	case nil:
+		return nil
+	case OperationError:
+		return err.OpError
+	default:
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:56]: I don't think the errors.As construct helps much here and it pulls in a module dependency. Plus I doubt we'll be adding lots of new error types to this so the As check also feels a bit of a YAGNI.

File: rados/omap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go:70]: I'm pretty sure it is a violation. Your first example passes a pointer to a C struct, and the second is about passing Go pointer to a C function, not about storing it in C memory, which is legal if the C function doesn't store it longer than it return. Here, however, we store a Go pointer in C memory for an indefinite time. The GC could decide any time to move the slice, and it will not correct the pointer to it, because it's in C memory, so it will become dangling and crash.

File: rados/read_op.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/read_op.go:56]: I'm confused. Which module are you talking about? We use `errors` all over the place. So it looks better, will not break if we might wrap errors in the future, and we are avoiding issues with linters (as just happened with CSI), and I really don't see any downsides. 🤷‍♂️

File: rados/omap.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go:24]: Oh, it does. `unsafe.Sizeof()` returns a constant expression, and I tried it before suggesting.

File: rados/operation.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/operation.go:29]: I can't agree it's a common go idiom. I  was curious and checked: https://play.golang.org/p/kJnVThxWTwv
`writeOp` is in fact of type `string` in this code.

File: rados/operation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/operation.go:145]: I think I don't understand. The numer of times `C.free()` is called is identical in both cases. I just didn't understand the reason for the temporary `r` variable and removed it. Am I missing something?

File: rados/omap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go:70]: From [Passing pointers](https://golang.org/cmd/cgo/#hdr-Passing_pointers)
> C code may not keep a copy of a Go pointer after the call returns.
[...]
Go code may not store a Go pointer in C memory. C code may store Go pointers in C memory, subject to the rule above: it must stop storing the Go pointer when the C function returns.

[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go:70]: And here is the proof:
<details><summary>(Click to expand)</summary>

<pre>
[svanders:~/golang/test] $ cat pointertest.go 
package main

// typedef void* ptr;
import "C"

import (
	"fmt"
)

func main() {
	a := []byte("This is a test")
	goPtr := &a[0]
	cPtr := (**byte)(C.malloc(C.sizeof_ptr))
	*cPtr = goPtr
	fmt.Printf("%p\n", cPtr)
}

[svanders:~/golang/test] $ GODEBUG=cgocheck=2 go run pointertest.go 
write of Go pointer 0xc0000b2010 to non-Go memory 0x65e820
fatal error: Go pointer stored into non-Go memory

runtime stack:
runtime.throw(0x4c57ce, 0x24)
	/usr/lib/golang/src/runtime/panic.go:1116 +0x72
runtime.cgoCheckWriteBarrier.func1()
	/usr/lib/golang/src/runtime/cgocheck.go:55 +0xa9
runtime.systemstack(0x462cb4)
	/usr/lib/golang/src/runtime/asm_amd64.s:370 +0x66
runtime.mstart()
	/usr/lib/golang/src/runtime/proc.go:1116

goroutine 1 [running]:
runtime.systemstack_switch()
	/usr/lib/golang/src/runtime/asm_amd64.s:330 fp=0xc00004ee20 sp=0xc00004ee18 pc=0x462de0
runtime.cgoCheckWriteBarrier(0x65e820, 0xc0000b2010)
	/usr/lib/golang/src/runtime/cgocheck.go:53 +0xcf fp=0xc00004ee58 sp=0xc00004ee20 pc=0x40588f
runtime.wbBufFlush(0x65e820, 0xc0000b2010)
	/usr/lib/golang/src/runtime/mwbbuf.go:196 +0x85 fp=0xc00004ee88 sp=0xc00004ee58 pc=0x42ee05
runtime.gcWriteBarrier(0x8, 0x65e820, 0x4431ca, 0x54c3e0, 0xc000000180, 0xc00004ef78, 0x406e45, 0xc00008e058, 0x0, 0x0, ...)
	/usr/lib/golang/src/runtime/asm_amd64.s:1461 +0xae fp=0xc00004ef10 sp=0xc00004ee88 pc=0x464c8e
main.main()
	/home/svanders/golang/test/pointertest.go:14 +0xf9 fp=0xc00004ef88 sp=0xc00004ef10 pc=0x49c3f9
runtime.main()
	/usr/lib/golang/src/runtime/proc.go:204 +0x209 fp=0xc00004efe0 sp=0xc00004ef88 pc=0x436289
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc00004efe8 sp=0xc00004efe0 pc=0x464bc1
exit status 2
</pre>
</details>

File: rados/operation.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:29]: Err, right. I got my wires crossed wrt to the "iota" use case (which I almost used):
```
const (
    foo opKind = "foo"
    bar
)
```
Does what I expected wrt the types, but is pretty useless for strings. I do find it interesting the other compile type checks didn't catch it. Regardless, good catch.

File: rados/omap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:70]: Got it. I was too fixated on the `unsafe.Pointer &b[0])` idiom, and not enough what we were *doing* with the result. Unfortunately, that means more copies in the code.
Since this is a metadata related op I think its probably fine to do copies for now. We can look into other techniques for performance in the future.

File: rados/operation.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Fixed.

File: rados/operation.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Fixed

File: rados/operation.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+func (e OperationError) Error() string {
+	subErrors := []string{}
+	if e.OpError != nil {
+		subErrors = append(subErrors,
+			fmt.Sprintf("op=%s", e.OpError.Error()))
+	}
+	for idx, es := range e.StepErrors {
+		subErrors = append(subErrors,
+			fmt.Sprintf("Step#%d=%s", idx+1, es.Error()))
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Fixed

File: rados/operation.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+	kind       opKind
+	OpError    error
+	StepErrors map[int]error
+}
+
+func (e OperationError) Error() string {
+	subErrors := []string{}
+	if e.OpError != nil {
+		subErrors = append(subErrors,
+			fmt.Sprintf("op=%s", e.OpError.Error()))
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Fixed

File: rados/operation.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:75]: Updated comments.

File: rados/operation.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:79]: updated comments

File: rados/operation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Since I can't explain myself, I reverted to the simpler form you suggested.

File: rados/operation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go]: Fixed

File: rados/omap.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Simplified.

File: rados/omap.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Changed to use CBytes *big sigh*

File: rados/omap.go (hunk 1 above)
Comment (phlogistonjohn) [reply rados/omap.go:126]: I don't think that's much clearer. Any strong objections to leaving it for now?

File: rados/omap.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Can we resolve as "something for another day"?

File: rados/omap.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Fixed.

File: rados/read_op.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:56]: I just really don't like relying on "new" language features unless its absolutely needed. I'm just very much in the habit of needing to support older language versions for other projects. Since its a bit of a wash which is clearer/simpler I figured why add the extra dependency (on the call - not just the package). I also don't think we'll be wrapping it any time soon. Do you mind if you solicit the other reviewers to get more opinions before deciding on this one?

File: rados/read_op.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:37]: Are you OK with this now?

File: rados/operation.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:94]: So it would be fully valid to clear o.steps here. However, I'm a bit reticent to do so because that would "separate" the steps from the op they belong to. Normally, that wouldn't be an issue but I am worried about cases where some debugging might be needed and if free has been called, the step objects could be hard to trace back to the op they were created for.

If this seems like a crappy justification for not adding a single line let me know, because I don't feel too strongly about it.

**Comment by phlogistonjohn** [comment]
I have pushed a new version of the patches taking into account @ansiwen review so far. Thanks!

File: rados/omap.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ansiwen) [reply rados/omap.go]: Sure

File: rados/omap.go (hunk 1 above)
Comment (ansiwen) [reply rados/omap.go:126]: No, I just found it very "technical", but not big issue, if you think that is better.

**Review by ansiwen — CHANGES_REQUESTED** [review]
Next round...

File: rados/operation.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+func (e OperationError) Error() string {
+	subErrors := []string{}
+	if e.OpError != nil {
+		subErrors = append(subErrors,
+			fmt.Sprintf("op=%s", e.OpError))
+	}
+	for idx, es := range e.StepErrors {
+		subErrors = append(subErrors,
+			fmt.Sprintf("Step#%d=%s", idx+1, es))
```
Comment (nixpanic) [inline rados/operation.go:49]: why start counting with 1, and not 0?

File: rados/operation.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+		o.steps[i].free()
+	}
+}
+
+// update the operation and the steps it contains. The top-level result
+// of the rados call is passed in as ret and used to construct errors.
+// The update call of each step is used to update the contents of each
+// step and gather any errors from those steps.
+func (o *operation) update(kind opKind, ret C.int) error {
+	stepErrors := make(map[int]error, 0)
```
Comment (nixpanic) [inline rados/operation.go:101]: is it needed to pass the initial size to `make` when zero?

File: rados/read_op.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nixpanic) [reply rados/read_op.go:37]: I'd like to see the check for a non-nil ioctx too, of course :wink:

File: rados/read_op.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (nixpanic) [reply rados/read_op.go:56]: I think the usage of the switch on error is fine here. Errors come from the same package, so I do not see the strong dependency as a problem. My opinion would be different if the error comes from an other package.

File: rados/operation.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+type withoutFree struct{}
+
+func (*withoutFree) free() {}
+
+// withRefs is a embeddable type to help track and free C memory.
+type withRefs struct {
+	refs []unsafe.Pointer
+}
+
+func (w *withRefs) free() {
```
[the remark this answers is not in the document]
Comment (nixpanic) [reply rados/operation.go:142]: I like `add()` more as well, but the addition of multiple-refs with a single `add()` makes it a little weird again. I am not sure if `add()` with a single pointer is sufficient?

File: rados/omap.go [hunk 20]
```diff
@@ -1,69 +1,101 @@
...
+	cLengths := C.malloc(maplen * sizeTSize)
+
+	sos := &setOmapStep{
+		pairs:    pairs,
+		cKeys:    (**C.char)(cKeys),
+		cValues:  (**C.char)(cValues),
+		cLengths: (*C.size_t)(cLengths),
+		cNum:     C.size_t(len(pairs)),
+	}
+	sos.addRefs(cKeys, cValues, cLengths)
```
Comment (nixpanic) [inline rados/omap.go:52]: in this context `addRefs()` makes sense, `add()` would not be self-explaining when reading the code. So maybe have `freeRefs()` instead of `free()`?

File: rados/omap.go [hunk 21]
```diff
@@ -1,69 +1,241 @@
...
-		C.size_t(len(pairs)))
+	runtime.SetFinalizer(sos, opStepFinalizer)
+	return sos
+}
+
+func (sos *setOmapStep) free() {
+	sos.cKeys = nil
+	sos.cValues = nil
+	sos.cLengths = nil
+	sos.withRefs.free()
```
[the remark this answers is not in the document]
Comment (nixpanic) [reply rados/omap.go:90]: No, `sos.free()` would be recursive? This looks like a great reason to call the function `freeRefs()`

File: rados/omap.go [hunk 22]
```diff
@@ -1,69 +1,241 @@
...
+// call will return an error. After Operate is called, the Next call will
+// return valid results.
+//
+// The life cycle of the GetOmapStep is bound to the ReadOp, if the ReadOp
+// Release method is called the public methods of the step must no longer be
+// used and may return errors.
+type GetOmapStep struct {
+	// inputs:
+	startAfter   string
+	filterPrefix string
```
[the remark this answers is not in the document]
Comment (nixpanic) [reply rados/omap.go:110]: maybe we should have some golang-ci job that identifies these unused attributes/functions/...

**Review by nixpanic — CHANGES_REQUESTED** [review]
Looks good to me in general, feel very clean. Just a few comments that need addressing or answers.

File: rados/operation.go (hunk 17 above)
Comment (phlogistonjohn) [reply rados/operation.go:49]: It seemed right at the time, but not I'm unsure. The goal is to be able to map back to operation steps that actually encountered an error. So, in the case that the entire error got printed/logged one could observe which operation failed. I think, at the time, I figured it would make more sense to use "human" numbering 1st step, 2nd step, etc. rather than "software" numbering. But now, I wonder how useful that'd be as it seems that requring the person reading the logs to know the exact order of the steps is a bit fishy.

Between this and @ansiwen other comments on this same type I am wondering if we ought to add ad name too. But we can do that later too. Opinions?

File: rados/operation.go (hunk 18 above)
Comment (phlogistonjohn) [reply rados/operation.go:101]: No, I can even drop the make.

File: rados/read_op.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:37]: I think I'd like to table this part of the PR discussion and make it a topic for the next time we talk. I think you'll all agree that we want to make calls in go-ceph consistent and so this part of things has wider reaching implications (IMO). If you can both stomach not checking it for now we can discuss the more "meta" aspects of this later and decide what approach we want throughout go-ceph.

File: rados/omap.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:110]: They're used in that they are assigned to. They're not read from after that. I had a a pattern of capturing the go inputs in each step even if things were converted to C types later. I kind of like the idea of hanging onto them for debgging purposes. I could add comments to that affect, or are they strong objections to keeping these references?

File: rados/omap.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:90]: Heh. So I agree that @nixpanic has noticed that the withRefs is needed to access the "right" free function. I had named it "free" as a short-cut for types that were not defining their own free function. In those cases the free function of the embedded withRefs type could satisfy the interface.
If you think that's too "cute"/confusing, I can change it to freeRefs.

File: rados/omap.go [hunk 23]
```diff
@@ -1,69 +1,241 @@
...
+// setOmapStep is a write op step. It holds C memory used in the operation.
+type setOmapStep struct {
+	withRefs
+	withoutUpdate
 
-	c_keys := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_values := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_lengths := C.malloc(C.size_t(len(pairs)) * C.size_t(unsafe.Sizeof(s)))
+	// user input
+	pairs map[string][]byte
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:29]: OK, as discussed in a voice call, the memory in the byte slices could be large and they are mutable so there is no guarantee they'll be the same as the copy in c. I will drop this from the struct.

File: rados/omap.go [hunk 24]
```diff
@@ -1,69 +1,241 @@
...
-		(*C.size_t)(c_lengths),
-		C.size_t(len(pairs)))
+	runtime.SetFinalizer(sos, opStepFinalizer)
+	return sos
+}
+
+func (sos *setOmapStep) free() {
+	sos.cKeys = nil
+	sos.cValues = nil
+	sos.cLengths = nil
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:89]: Just to re-use the withRefs mechanism as much as possible.

File: rados/omap.go [hunk 25]
```diff
@@ -1,69 +1,241 @@
...
+	keys []string
+
+	// arguments:
+	cKeys **C.char
+	cNum  C.size_t
+}
+
+func newRemoveOmapKeysStep(keys []string) *removeOmapKeysStep {
+	var c *C.char
+	ptrSize := unsafe.Sizeof(c)
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:207]: Will fix.

File: rados/omap.go [hunk 26]
```diff
@@ -1,69 +1,241 @@
...
+}
+
+// removeOmapKeysStep is a write operation step used to track state, especially
+// C memory, across the setup and use of a WriteOp.
+type removeOmapKeysStep struct {
+	withRefs
+	withoutUpdate
+
+	// inputs:
+	keys []string
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:198]: Sure.

File: rados/operation.go [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	for idx, es := range e.StepErrors {
+		subErrors = append(subErrors,
+			fmt.Sprintf("Step#%d=%s", idx+1, es))
+	}
+	return fmt.Sprintf(
+		"%s operation error: %s",
+		e.kind,
+		strings.Join(subErrors, ", "))
+}
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:55]: I did see the suggestion. We dIscussed this in a voice call and decided it was OK due to the  more specific error output.

File: rados/omap.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:110]: These are short strings that don't take up a lot of resources and can be useful while debugging as these are native go types versus the C conversions we actually pass to ceph.

File: rados/read_op.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/read_op.go:37]: FWIW I have added a test to ensure passing a nil ioctx panics.

File: rados/omap.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Field revmoed.

File: rados/omap.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Updated to use a single ptrSize in the file.

File: rados/omap.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go]: Field removed.

File: rados/omap.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:105]: Keeping these as per verbal discussion.

File: rados/omap.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/omap.go:85]: Keeping, as it helps simpler code using an embedded withRefs automatically meet the interface.

File: rados/operation.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rados/operation.go:142]: Changed to 'add' with a single argument.

File: rados/operation.go (hunk 17 above)
Comment (phlogistonjohn) [reply rados/operation.go]: Changed to use use zero indexing.

File: rados/operation.go (hunk 18 above)
Comment (phlogistonjohn) [reply rados/operation.go]: Dropped.

File: rados/omap.go (hunk 20 above)
Comment (phlogistonjohn) [reply rados/omap.go]: I went with add/free as per the other conversation. Please reopen this subthread if you think this needs addtional discussion.

**Comment by phlogistonjohn** [comment]
I have updated the PR to take the previously outstanding feedback. I closed as many of the subtreads as I could so that it was clear that feedback had been acted on or considered. If a topic needs to be revisited, feel free, I just didn't want lots of stale sub-threads hanging open.

PTAL @ansiwen @nixpanic

**Review by ansiwen — APPROVED** [review]
LGTM. :tada: :smiley:

**Review by nixpanic — APPROVED** [review]
